src/backend/distributed/catalog/dependency.c

```diff
@@ -0,0 +1,381 @@
+ *  - the pg_depend entry is a normal dependency, all other types are created and
+ *    maintained by postgres
+ *  - the object is not already in the targetList (context), if it is already in there an
+ *    other object already caused the creation of this object
+ *  - the object is not already marked as distributed in pg_dist_object. Objects in
+ *    pg_dist_object are already available on all workers.
+ *  - object is not created by an other extension. Objects created by extensions are
+ *    assumed to be created on the worker when the extension is created there.
+ *  - citus supports object distribution. If citus does not support the distribution of
+ *    the object we will not try and distribute it.
+ */
+static bool
+follow_get_dependencies_for_object(void *context, const Form_pg_depend pg_depend)
+{
+	List **targetList = (List **) context;
+	ObjectAddress address = { 0 };
+	ObjectAddressSet(address, pg_depend->refclassid, pg_depend->refobjid);
+
+	if (pg_depend->deptype != DEPENDENCY_NORMAL)
+	{
+		return false;
+	}
+
+	if (IsObjectAddressInList(&address, *targetList))
+	{
+		return false;
+	}
+
+	if (isObjectDistributed(&address))
+	{
+		return false;
+	}
+
+	if (IsObjectAddressOwnedByExtension(&address))
+	{
+		return false;
+	}
+
+	if (!SupportedDependencyByCitus(&address))
+	{
+		return false;
+	}
+
+	return true;
+}
+
+
+/*
+ * follow_order_object_address applies filters on pg_depend entries to follow the
+ * dependency tree of objects in depth first order. The filters are practically the same
+ * to follow_get_dependencies_for_object, except it will follow objects that have already
+ * been distributed. This is because its primary use is to order the objects from
+ * pg_dist_object in dependency order.
+ *
+ * For completeness the list of all edges it will follow;
+ *
+ * Objects are only followed if all of the following checks hold true:
+ *  - the pg_depend entry is a normal dependency, all other types are created and
+ *    maintained by postgres
+ *  - the object is not already in the targetList (context), if it is already in there an
+ *    other object already caused the creation of this object
+ *  - object is not created by an other extension. Objects created by extensions are
+ *    assumed to be created on the worker when the extension is created there.
+ *  - citus supports object distribution. If citus does not support the distribution of
+ *    the object we will not try and distribute it.
+ */
+static bool
+follow_order_object_address(void *context, const Form_pg_depend pg_depend)
+{
+	List **targetList = (List **) context;
+	ObjectAddress address = { 0 };
+	ObjectAddressSet(address, pg_depend->refclassid, pg_depend->refobjid);
+
+	if (pg_depend->deptype != DEPENDENCY_NORMAL)
+	{
+		return false;
+	}
+
+	if (IsObjectAddressInList(&address, *targetList))
+	{
+		return false;
+	}
+
+	if (IsObjectAddressOwnedByExtension(&address))
+	{
+		return false;
+	}
+
+	if (!SupportedDependencyByCitus(&address))
```

Comment:
Could maybe move this check above IsObjectAddressInList since it's relatively cheap.

Reply:
for now yes, as stated on the types PR, I have been thinking about providing warnings around unsupported objects.

I think if we do so now we will actually have too many warnings, hence I didn't do so. For now I will move the check up.